fix: update and add quality tests

samuelallan72 opened the pull request:
## Description

- update requirements
- update pylint from edx-lint
- configure quality tests and linters: pylint, isort, black
- fix the code to pass the lints
- update .github files

## Supporting information

Private-ref: https://tasks.opencraft.com/browse/BB-10035

## Test instructions

- verify the CI passes
- clone locally and run `tox -e quality`
- also test the api endpoints to verify behaviour hasn't changed - see #1 and #2 

## Deadline

None

samuelallan72 commented on pylintrc_tweaks:
```diff
@@ -0,0 +1,9 @@
+    missing-function-docstring,
+    consider-using-f-string,
+    too-few-public-methods,
+    django-not-configured,
```
I added these because they're low priority and many of them. The import errors are about all the edx-platform imports that aren't available in the tox environment.

When I add some integration or unit tests in a future PR, those imports will be tested.

Kelketek commented:
@samuelallan72 I've done a readthrough and these changes make sense to me. I may circle back to test the endpoints manually after some other work today, but I'm just as happy to let @tecoholic test it since I think he's already got it set up.

samuelallan72 committed:
feat: update dependencies, tidy repo files, add CI

samuelallan72 commented:
@tecoholic the diff is fairly large, and I was worried about creating subtle bugs by mixing refactoring with reformatting. So I split this into 3 commits, which may make it easier to review separately:

1. changes to all the scaffolding files
2. reformatting (this is only the result of running `make format`)
3. refactoring and tweaks to fix quality checks

A few minor things changed in the process (nothing functional) - you can check those with the "compare" button next to the force push message above.

cc @Kelketek

tecoholic commented on pylintrc:
```diff
@@ -355,10 +288,9 @@ no-docstring-rgx = __.*__$|test_.+|setUp$|setUpClass$|tearDown$|tearDownClass$|M
 
 [FORMAT]
-max-line-length = 100
+max-line-length = 120
```
@samuelallan72 We have the line length being set to 120 here. But this config isn't shared with Black. So black uses the default for `88` for formatting.

Black seems to load the config [only from "pyproject.toml"](https://black.readthedocs.io/en/stable/usage_and_configuration/the_basics.html#configuration-via-a-file) at the project level. We should either add that file and set the `line-length` to `120` for consistency or at least pass it as an CLI argument in the Makefile `black -l 120`. But later is sub-optimal.

Can we add a pyproject.toml and put Black configuration there? I think that would actually make this diff a lot smaller than now.

tecoholic replied:
Maybe not a lot smaller, but really reduce the code formatting related changes.

samuelallan72 replied:
Ah nice catch, let me do that now. :+1:

samuelallan72 replied:
I changed the line length, reformatted, and re-sorted the commits. I also added a commit https://github.com/open-craft/legacy-appsembler-api/pull/3/commits/6a605788308ab75e78fa3f5714550700520c2ed9 - this is something I had added previously, but lost the change when I split the commits last time.

Also, sorry I didn't realise you were still reviewing, so I haven't addressed your other comments.

tecoholic replied:
@samuelallan72 Thank you for addressing this. I think I clicked Add comment instead of "Review" for this comment and it got published before the rest of the comments. I am sorry if it caused confusion for you.

samuelallan72 replied:
All good :grin:

tecoholic commented on setup.cfg:
```diff
@@ -1,8 +1,5 @@
 [isort]
```
Related to the comment on Black & line legth -  we can replace this `setup.cfg` with pyproject.toml.

tecoholic commented on setup.py:
```diff
@@ -18,12 +18,12 @@ def get_version(*file_paths):
-    version_file = open(filename).read()
-    version_match = re.search(r"^__version__ = ['\"]([^'\"]*)['\"]",
-                              version_file, re.M)
+    with open(filename, encoding="utf-8") as version_file_handle:
```
Python 3 by default uses the "utf-8" encoding right? More or a curious question, why change from simple open() to specifying the encoding?

samuelallan72 replied:
I took a look, and it's not actually the default:

> In text mode, if encoding is not specified the encoding used is platform-dependent: [locale.getencoding()](https://docs.python.org/3/library/locale.html#locale.getencoding) is called to get the current locale encoding.

source: https://docs.python.org/3/library/functions.html#open

Also, pylint was complaining about the encoding being missing - that's the main reason I changed it there. :)

I did also find [PEP 686](https://peps.python.org/pep-0686/) which has been accepted, which will make utf-8 default from python 3.15 though.

tecoholic replied:
@samuelallan72 Oh! wow. This is a surprise. Thank you for the docs. After so many years, it's amazing how we still learn new things about basic stuff.

samuelallan72 replied:
No worries, I had to double check myself! There's a lot out there. :)

tecoholic commented on setup.py:
```diff
@@ -93,49 +104,53 @@ def is_requirement(line):
+        "License :: Other/Proprietary License",
+        "Natural Language :: English",
+        "Programming Language :: Python :: 3",
+        "Programming Language :: Python :: 3.5",
```
```suggestion
        "Programming Language :: Python :: 3.11",
```

tecoholic commented on setup.py:
```diff
@@ -93,49 +104,53 @@ def is_requirement(line):
-        'Programming Language :: Python :: 3.5',
+        "Development Status :: 3 - Alpha",
+        "Framework :: Django",
+        "Framework :: Django :: 2.2",
```
```suggestion
        "Framework :: Django :: 4.2",
```
Sumac is on Django 4.2

samuelallan72 replied:
Ah yeah I haven't touched the metadata in the setup.py. Maybe we can address that later if you think it's important? We're not publishing to pypi, so it's probably fine to leave as-is.

tecoholic commented on shoppingcart/migrations/0005_new_initial.py:
```diff
@@ -1,45 +1,97 @@
 # Generated by Django 4.2.20 on 2025-09-19 05:43
```
I think we can add migrations to the "excluded" things in Black. Being a machine generated file, there is no benefit to formatting it with a formatter. Also, regnerating them for minor changes will create very noisy diffs.

samuelallan72 replied:
That file is actually partially manually created - I needed to edit it for squashing. Personally I'd prefer to keep even these files included for the formatter - I'd don't see how it will make noisy diffs? The files will be created or edited, then run through the formatter, then committed and pushed - there isn't really any churn.

tecoholic replied:
> That file is actually partially manually created

Ah! Then it makes sense to prefer formatting. 

> I'd don't see how it will make noisy diffs?

Now that I think about it, it's mostly solved at the PR review time. I run into noisy diff situations when working on a new feature. I would add an attribute and create a migration, then adjust something and recreate the same migration, and my editor or git diff would show everything has changed in the file, instead of single line that I would like to see. Think of fixing the spelling of something for example.

Anyway, I guess I have built it into "machine generated means don't change" thumb rule.

samuelallan72 committed:
feat: format with isort and black

tecoholic commented on shoppingcart/models.py:
```diff
@@ -59,17 +74,23 @@ def get_registration_code_redemption(cls, code, course_id):
-    def create_invoice_generated_registration_redemption(cls, course_reg_code, user):  # pylint: disable=invalid-name
+    def create_invoice_generated_registration_redemption(
+        cls, course_reg_code, user
+    ):  # pylint: disable=invalid-name
```
I was very curious about the 'invalid-name' here. Turns out there is a 40 char limit on function names. I didn't know that was a Pylint rule.

samuelallan72 committed:
fix: fix failing quality checks

tecoholic reviewed (changes requested):
@samuelallan72 Amazing work on getting this repo to shape. If we sort out Black's config to match Pylint's rules on line-legth, I think we are good here.

tecoholic commented:
@samuelallan72 Oh! Thank you for the 3 clean commits. It was a breeze to review the PR.

samuelallan72 committed:
fix: use secure random generator for credentials

Random passwords and enrolment tokens are generated here.
The `random` module is a pseudo random number generator,
and not suitable for generating secure secrets.
Instead we can use the `secrets` module as a drop in replacement here.

samuelallan72 commented:
Thanks @tecoholic ! :)

I addressed your comment about the line length, and responded to your other comments. Not sure if you'd like any of your other comments addressed here before merge?

I also added a commit to fix a minor security issue - [6a60578](https://github.com/open-craft/legacy-appsembler-api/pull/3/commits/6a605788308ab75e78fa3f5714550700520c2ed9)

I did force push sorry, just to keep the commits all neat still. I guess next time I should submit them as separate PRs to avoid that.

This should be ready for another pass. :)

tecoholic reviewed (approved):
@samuelallan72 :+1: LGTM

- [x] I tested this: Ran the quality checks and formatting locally.
- [x] I read through the code
- [ ] I checked for accessibility issues - NA
- [ ] Includes documentation
